apollo-router/src/services/subgraph_service.rs

```diff
@@ -628,9 +629,54 @@ async fn call_websocket(
         }
         _ => connect_async(request).instrument(subgraph_req_span).await,
     }
-    .map_err(|err| FetchError::SubrequestWsError {
-        service: service_name.clone(),
-        reason: format!("cannot connect websocket to subgraph: {err}"),
+    .map_err(|err| {
+        let error_details = match &err {
+            tokio_tungstenite::tungstenite::Error::Utf8 => {
+                if let Some(utf8_err) = err
+                    .source()
+                    .and_then(|e| e.downcast_ref::<std::str::Utf8Error>())
```

Comment:
Since the  `Utf8` branch holds no data, i don't think that `.source()` can ever return `Some()` here..? am I missing something?

Reply:
I think it returns Some if it's the right type. It doesn't care about the data

Reply:
I meant `err.source()`, not the downcast ^^

Reply:
I don't think it changes anything if Utf8 variant doesn't hold data

Reply:
`Error::Utf8` is constructed in this file: https://github.com/snapview/tungstenite-rs/blob/8cfd027674a53620f08b6ad3af255f5638eb2f51/src/protocol/message.rs#L30
there's not a `std::str::Utf8Error` involved at any point, and `Error::Utf8` does not carry a source error

Reply:
I got your point sorry I misread. I think you're right. I'll take a deeper look at this